ddtrace/bootstrap/sitecustomize.py

```diff
@@ -2,30 +2,112 @@
 
-    gevent.monkey.patch_all()
 
+import os  # noqa
+
+
+MODULES_TO_NOT_CLEANUP = {"atexit", "asyncio", "attr", "concurrent", "ddtrace", "logging"}
+if sys.version_info <= (2, 7):
+    MODULES_TO_NOT_CLEANUP |= {"encodings", "codecs"}
+    import imp
+
+    _unloaded_modules = []
+else:
+    import importlib
+
+
+def is_installed(module_name):
+    # https://stackoverflow.com/a/51491863/735204
+    if sys.version_info >= (3, 4):
+        return importlib.util.find_spec(module_name)
+    elif sys.version_info >= (3, 3):
+        return importlib.find_loader(module_name)
+    elif sys.version_info >= (3, 1):
+        return importlib.find_module(module_name)
+    elif sys.version_info >= (2, 7):
+        try:
+            imp.find_module(module_name)
+        except ImportError:
+            return False
+        else:
+            return True
+    return False
+
+
+def should_cleanup_loaded_modules():
+    dd_unload_sitecustomize_modules = os.getenv("DD_UNLOAD_MODULES_FROM_SITECUSTOMIZE", default="0").lower()
+    if dd_unload_sitecustomize_modules == "0":
+        return False
+    elif dd_unload_sitecustomize_modules not in ("1", "auto"):
+        return False
+    elif dd_unload_sitecustomize_modules == "auto" and not any(
+        is_installed(module_name) for module_name in MODULES_THAT_TRIGGER_CLEANUP_WHEN_INSTALLED
+    ):
+        return False
+    return True
+
+
+def cleanup_loaded_modules(aggressive=False):
+    """
+    "Aggressive" here means "cleanup absolutely every module that has been loaded since startup".
+    Non-aggressive cleanup entails leaving untouched certain modules
+    This distinction is necessary because this function is used both to prepare for gevent monkeypatching
+    (requiring aggressive cleanup) and to implement "module cloning" (requiring non-aggressive cleanup)
+    """
+    modules_loaded_since_startup = set(_ for _ in sys.modules if _ not in MODULES_LOADED_AT_STARTUP)
+    # Figuring out modules_loaded_since_startup is necessary because sys.modules has more in it than just what's in
+    # import statements in this file, and unloading some of them can break the interpreter.
+    modules_to_cleanup = modules_loaded_since_startup - MODULES_TO_NOT_CLEANUP
+    if aggressive:
+        modules_to_cleanup = modules_loaded_since_startup
+    # Unload all the modules that we have imported, except for ddtrace and a few
+    # others that don't like being cloned.
+    # Doing so will allow ddtrace to continue using its local references to modules unpatched by
+    # gevent, while avoiding conflicts with user-application code potentially running
+    # `gevent.monkey.patch_all()` and thus gevent-patched versions of the same modules.
+    for m in modules_to_cleanup:
+        if not aggressive and any(
+            m.startswith("%s." % module_to_not_cleanup) for module_to_not_cleanup in MODULES_TO_NOT_CLEANUP
+        ):
+            continue
+        del sys.modules[m]
+
+    if "time" in sys.modules:
```

Comment:
why? comment could be nice

Reply:
From what i recall, the time module is not always guaranteed to be in `sys.modules` when we initialize `MODULES_LOADED_AT_STARTUP`, and this causes issues if we don't clone it on ddtrace's end as well. @P403n1x87 can you confirm?